lib/caching/inMemoryLock.ts

```diff
@@ -0,0 +1,45 @@
+import log from 'loglevel';
+export class InMemoryLock {
+  private locks: { [key: string]: { isLocked: boolean; requestQueue: (() => void)[] } };
+
+  constructor() {
+    this.locks = {};
+  }
+
+  lock(key: string): Promise<void> {
+    return new Promise(resolve => {
+      if (!this.locks[key]) {
+        this.locks[key] = { isLocked: false, requestQueue: [] };
+      }
+
+      const request = () => {
+        log.debug(`[InMemoryLock] `, new Date(), ` - Locking: ${key}. `, new Date());
+        this.locks[key].isLocked = true;
+        resolve();
+      };
+
+      if (this.locks[key].isLocked) {
+        log.debug(`[InMemoryLock] `, new Date(), ` - ${key} added to lock Queue. `, new Date());
+        this.locks[key].requestQueue.push(request);
+      } else {
+        request();
+      }
+    });
+  }
+
+  unlock(key: string) {
+    log.debug(`[InMemoryLock] `, new Date(), ` - Unlocking ${key}. `, new Date());
+
+    const lock = this.locks[key];
+    if (!lock) {
+      throw new Error(`No lock found for key: ${key}`);
+    }
+
+    lock.isLocked = false;
```

Comment:
Should this variable be `let lock`, can you change the `isLocked` property?

Reply:
a `const` will only stop you from replace the whole object. but changing individual properties is ok